kythe/cxx/indexer/cxx/IndexerASTHooks.cc

```diff
@@ -3827,856 +3992,606 @@ void IndexerASTVisitor::DumpTypeContext(unsigned Depth, unsigned Index) {
+    clang::FunctionProtoTypeLoc TL) {
+  std::vector<GraphObserver::NodeId> NodeIds;
+  std::vector<const GraphObserver::NodeId*> NodeIdPtrs;
+  auto ReturnType = BuildNodeIdForType(TL.getReturnLoc());
+  if (!ReturnType) {
+    return NodeSet::Empty();
+  }
+  NodeIds.push_back(*ReturnType);
+
+  for (const auto& param : TL.getTypePtr()->getParamTypes()) {
+    if (auto ParmType = BuildNodeIdForType(param)) {
+      NodeIds.push_back(*ParmType);
+    } else {
+      return NodeSet::Empty();
+    }
+  }
+
+  for (const auto& node : NodeIds) {
+    NodeIdPtrs.push_back(&node);
+  }
+  const char* Tycon = TL.getTypePtr()->isVariadic() ? "fnvararg" : "fn";
+  return Observer.recordTappNode(Observer.getNodeIdForBuiltinType(Tycon),
+                                 NodeIdPtrs, NodeIdPtrs.size());
+}
+
+NodeSet IndexerASTVisitor::BuildNodeSetForFunctionNoProto(
+    clang::FunctionNoProtoTypeLoc TL) {
+  return Observer.getNodeIdForBuiltinType("knrfn");
+}
+
+NodeSet IndexerASTVisitor::BuildNodeSetForParen(clang::ParenTypeLoc TL) {
+  return BuildNodeSetForType(TL.getInnerLoc());
+}
+
+NodeSet IndexerASTVisitor::BuildNodeSetForDecltype(clang::DecltypeTypeLoc TL) {
+  return BuildNodeSetForType(TL.getTypePtr()->getUnderlyingType());
+}
+
+NodeSet IndexerASTVisitor::BuildNodeSetForElaborated(
+    clang::ElaboratedTypeLoc TL) {
+  return BuildNodeSetForType(TL.getNamedTypeLoc());
+}
+
+NodeSet IndexerASTVisitor::BuildNodeSetForTypedef(clang::TypedefTypeLoc TL) {
+  // TODO(zarko): Return canonicalized versions as well.
+  GraphObserver::NameId AliasID = BuildNameIdForDecl(TL.getTypedefNameDecl());
+  // We're retrieving the type of an alias here, so we shouldn't thread
+  // through the deduced type.
+  if (auto AliasedTypeID = BuildNodeIdForType(
+          TL.getTypedefNameDecl()->getTypeSourceInfo()->getTypeLoc())) {
+    // TODO(shahms): Move caching here and use
+    // `Observer.nodeIdForTypeAliasNode()` when already built?
+    // Or is always using the cached id sufficient?
+    NodeId ID = Observer.nodeIdForTypeAliasNode(AliasID, *AliasedTypeID);
+    auto Marks = MarkedSources.Generate(TL.getTypedefNameDecl());
+    return Observer.recordTypeAliasNode(
+        ID, *AliasedTypeID,
+        BuildNodeIdForType(FollowAliasChain(TL.getTypedefNameDecl())),
+        Marks.GenerateMarkedSource(ID));
+  }
+  return NodeSet::Empty();
+}
+
+NodeSet IndexerASTVisitor::BuildNodeSetForSubstTemplateTypeParm(
+    clang::SubstTemplateTypeParmTypeLoc TL) {
+  return BuildNodeSetForType(TL.getTypePtr()->getReplacementType());
+}
+
+NodeSet IndexerASTVisitor::BuildNodeSetForInjectedClassName(
+    clang::InjectedClassNameTypeLoc TL) {
+  // TODO(zarko): Replace with logic that uses InjectedType.
+  return BuildNodeSetForNonSpecializedRecordDecl(TL.getDecl());
+}
+
+NodeSet IndexerASTVisitor::BuildNodeSetForDependentName(
+    clang::DependentNameTypeLoc TL) {
+  // TODO(shahms): EmitRanges below should be No.
```

Comment:
why isn't it?

Reply:
Removed.

Reply:
Because we fail to emit relevant edges; comment updated.